src/components/shared/thread/ActivityGraphFills.js

```diff
@@ -478,18 +497,66 @@ export class ActivityFillGraphQuerier {
     // (In fact, yPercentage > upperEdgeOfPreviousSample except during the first
     // iteration - in the first iteration, yPercentage can be == categoryLowerEdge.)
     for (const { sample, contribution } of candidateSamples) {
+      const stackIndex = samples.stack[sample];
+      const sampleCategory =
+        stackIndex !== null
+          ? stackTable.category[stackIndex]
+          : greyCategoryIndex;
       const upperEdgeOfThisSample = upperEdgeOfPreviousSample + contribution;
-      if (yPercentage <= upperEdgeOfThisSample) {
+      if (sampleCategory === category && yPercentage <= upperEdgeOfThisSample) {
```

Comment:
I added a comment. Unfortunately it's not enough because there can be different samples in different categories that their yPercentage is less than upperEdgeOfThisSample. It's possible to pick the wrong one.